### ISSUE TITLE
Provide a way to select elements to monitor

### ISSUE THREAD

**Issue by javydreamercsw** [retest/recheck-web#326]
I'm loving the tool but sometimes feels like is trying to do too much leading to too much noise.

It would be awesome if we could annotate WebElements within a PageObject to be monitored (meaning screenshots taken as it's being interacted with) ignoring the to non-monitored ones.

At work there are multiple regions working on different parts of the complex application that we don't care about in some focused tests.

Looks like this goes against #29. In our case the current behavior generates way too many files already.

**Comment by roesslerj** [comment]
Hey! Thanks for the feedback. We're glad you like it. 😃

We're not sure we understood what your challenge is. Could we maybe have a short session via Skype, hangout, Jitsi, ... were you show us your problem? Then it'd be much easier for us to conceive a solution strategy.

Thanks!

**Comment by diba1013** [comment]
If I understand you correctly, you want to focus only on specific `WebElements` (i.e. `<div></div>`, etc.) while ignoring the noise inside the GUI.

### Current way

An alternative, that is **currently** available, is to define a custom filter or ignore-rules (which we currently have no documentation for!?), but essentially boils down to [this](https://docs.retest.de/recheck/how-ignore-works/).

A ignore can be defined in `${project-root}/.retest/recheck.ignore` with the [rules](https://docs.retest.de/recheck/how-ignore-works/) mentioned above. There you can specify globally for all tests, which elements you do not want to monitor.

If you want to have this only for a single test (advanced), we are currently working on #182 which is partly implemented for [suites ](https://github.com/retest/recheck/pull/193) while we wait for the support for [steps](https://github.com/retest/recheck-web/issues/182). See [this comment](https://github.com/retest/recheck-web/issues/182#issuecomment-488715207) on how it should work.

A filter can be defined in `${project-root}/.retest/filter/${name-of-filter}.filter` which is displayed in **_review_** or can be referenced inside **_recheck.cli_** while diffing. It is a way to reduce the noise within the GUI and temporarily removing all differences for which the filter applies. Thus here you can define&mdash;similarly to the ignore-mechanism&mdash;which regions to ignore.

_We are happy to help and to extend our documentation as needed._

### Proposal

A simple implementation would be to provide a [RecheckAdapter](https://github.com/retest/recheck/blob/master/src/main/java/de/retest/recheck/RecheckAdapter.java) that focuses on a `WebElement` instead of a `WebDriver`. This adapter would access the element via JS, essentially executing [getAllElementsByPath#transform](https://github.com/retest/recheck-web/blob/076c4b73bdfacb82be1b9f017f23719a500d161f/src/main/resources/javascript/getAllElementsByPath.js#L69) and finally converting the element similar to [PeerConverter#convertToPeer](https://github.com/retest/recheck-web/blob/277c284b6a7fe88b05c1c4dfe7ba068d1acc4ad9/src/main/java/de/retest/web/PeerConverter.java#L57) to a `RootElement` or multiple, if it is a `List<WebElement>` (preferably through two different adapters?).

The [UnbreakableDriver](https://github.com/retest/recheck-web/blob/cca4da8df6340936fd8b28220cfee5c60bee11be/src/main/java/de/retest/web/selenium/UnbreakableDriver.java) should take care of changing elements?

Usage proposal:
```
re.startTest("foo");

driver.get( url );

List<WebElement> buttons = driver.findElement( By.tagName("button") );
re.check( buttons, "buttons" );

WebElement profile = driver.findElement( By.id("profile") );
re.check( profile , "profile" );

re.capTest();
```

**Comment by javydreamercsw** [comment]
The proposal sounds like a good first step.

What I had in mind involved the Page Object pattern. Where you have WebElement annotated by @FindBy. Would be great if there was a way to do this monitoring just by adding an annotation to the WebElement.

But with your proposal this could be easily implemented. Most likely would require a Page Object class to extend from and an annotation analyzer to feed into recheck

**Comment by javydreamercsw** [comment]
The filters sound like a great idea at all those levels but seems really complicated specially on a complex application. It will be easier to list what you want vs. what you don't.

Maybe is my lack of understanding of the filter format but even at the step level sounds like way too much to handle in a big code base and might be way too complicated for the test developers to handle.

**Comment by diba1013** [comment]
That Design Pattern for [Page Object](https://github.com/SeleniumHQ/selenium/wiki/PageObjects) seems like a decend way combined with our approach to be able to test a website.

-  You still could use the driver and test the whole page
-  You could use the  Page Object to only test the Elements specified

However, what i gathered from my reading about Page Objects, you can use any arbitrary class as a Page. We therefore&mdash;as you specified&mdash;need to be able to identify these classes.

### Updated Proposal

1. Define a specific interface (`PageObject`) or annotation (`@PageObject`) to identify such a object as using reflection and annotation analyzer seems way to expensive for trying to identify an arbitrary object as a Page Object.
2. Implement a RecheckAdapter for each `WebElement` and `List<WebElement>` (see first proposal).
3. Implement a RecheckAdapter for scanning the `PageObject` (as annotated above) and scan all fields for any of `@FindBy`, `@FindBys` and `@FindAll` being either a `WebElement` or a `List<WebElement>`, essentially converting those into a single `List<WebElement>` for the above adapters to convert.

```{java}
// LoginPage.java
@PageObject
public class LoginPage {

	@FindBy( id = "btn-login" )
	private WebElement login;

	public LoginPage( final WebDriver driver ) {
		PageFactory.initElements( driver, this );
	}

	public void login() {
		login.click();
	}
}

// LoginTest.java
@Test
void login() {
	re.startTest();

	driver.get( "https://example.com" );
        re.check( driver, "site"); // For full site

	final LoginPage loginPage = new LoginPage( driver );
	re.check( loginPage, "open" ); // For a subset of elements

	loginPage.login();
	re.check( loginPage, "login" );

	re.capTest();
}
```

_Please tell me, if I missed something._ 

Since we have only recently started working with selenium, we rely on more experienced testers to inform us what the state-of-the-art testing is. As such: **Thank you for your insight**.

**Comment by javydreamercsw** [comment]
This sounds and looks perfect! Still I fear this won't be enough. I use various page objects to move through the application to get to the actual purpose of the test in question. Depending on the scope of the test a page object might be of interest in one test and not others. Still will need to switch monitoring on and off as needed. 

Feel free to reach out if you need more feedback.

**Comment by beatngu13** [comment]
>The filters sound like a great idea at all those levels but seems really complicated specially on a complex application. It will be easier to list what you want vs. what you don't.

Our basic mantra is: Start with checking everything and then ignore those things you don't care about, rather than the other way around. (See ["Whitelist Testing vs. Blacklist Testing"](https://medium.com/@roesslerj/whitelist-testing-vs-blacklist-testing-cac5b9435aa1)).

However, I like the idea of allowing to pass both to recheck-web; a `WebDriver` or a `WebElement`. But as @roesslerj said, a short session would be helpful to better understand the problem and challenges.

**Comment by javydreamercsw** [comment]
As long as there's a way to ignore on demand, via code I'm fine.

If you a list of elements that should be enough.

**Comment by javydreamercsw** [comment]
I'm open for a session sometime during the week if you guys are available.

**Comment by roesslerj** [comment]
We currently plan to have a release next week, where we create a beta-jar containing the necessary changes. Would you be willing to test that and give us feedback on it?

**Comment by javydreamercsw** [comment]
Sure do. I've been anxiously waiting for it.

**Comment by javydreamercsw** [comment]
Any updates?

**Comment by beatngu13** [comment]
@javydreamercsw we've got you covered: #338. This will be part of the next (v1.5.0) release. We'll keep updating this issue as soon as we have moved on with the PR.

**Comment by javydreamercsw** [comment]
Just some questions that came to mind while I'm preparing a demo including Recheck:

1. Is there a way to get from the assertion failure the current image and the baseline image it's being compared against?
1. With the proposed changes above will you get access to the specific element that had the difference and the ability to retrieve the baseline and diff of the comparison?

**Comment by roesslerj** [comment]
When the test fails, there is a report being generated. That report contains exactly those differences and the corresponding images. Have you tried review (the GUI)? Using it just for viewing the differences is free, only for maintaining the Golden Master you need a plan.
https://retest.de/review/
If you want, we can have a call, where I show you a bit about it...

**Comment by javydreamercsw** [comment]
I tried but the results I see in output doesn't match the report when I open it in the review app.

For example I got:
```
1 check(s) in 'com.pingidentity.p2.examples.visual_inspection.RecheckTest' found the following difference(s):
Test 'Example-Test' has 15 difference(s) in 1 state(s)
```

But the GUI shows this:
![Screen Shot 2019-09-11 at 4 15 25 PM](https://user-images.githubusercontent.com/2817343/64736045-63c91a80-d4af-11e9-8160-31779b800b2d.png)

**Comment by javydreamercsw** [comment]
I'm available for a call.

**Comment by javydreamercsw** [comment]
Now it worked. Had to clean and rerun the test

**Comment by roesslerj** [comment]
Puh. Have to admit you made me a bit nervous there. You mean a call now? Can you join that: https://meet.jit.si/retest

**Comment by roesslerj** [comment]
Or suggest a day and time that works for you :-)

**Comment by javydreamercsw** [comment]
Started giving it a try and noticed that the AutocheckingRecheckDriver constructor changed. There seems to no longer be a way to set a naming strategy. In my case I used to pass a custom AutocheckingCheckNamingStrategy and that's no longer possible it seems.

**Comment by javydreamercsw** [comment]
Tried using something like this:
```
new AutocheckingRecheckDriver(
                    (RemoteWebDriver) webDriver,
                    RecheckOptions.builder()
                        .namingStrategy(new CustomCheckNamingStrategy())
                        .build()));
```
When it used to work before like this:
```
new AutocheckingRecheckDriver(
                    (RemoteWebDriver) webDriver,
                    RecheckOptions.builder().build(),
                    new CustomCheckNamingStrategy()));
```

**Comment by javydreamercsw** [comment]
Finally figured it out:
```
new AutocheckingRecheckDriver(
                    (RemoteWebDriver) webDriver,
                    RecheckWebOptions.builder()
                        .checkNamingStrategy(new CustomCheckNamingStrategy())
                        .build()));
```

I'll keep digging and provide feedback ASAP.

**Comment by javydreamercsw** [comment]
Here's some initial feedback:

1. For some reason I was expecting the screenshots to be only of the element I'm checking and not the whole page.
1. I was unable to pass a list of elements and/or the PageObject itself to check more than one element.

**Comment by javydreamercsw** [comment]
Also it seems to only work with RecheckImpl() and not with RecheckWebImpl(). Took me a while to figure out why it didn't work on my project at work.

**Comment by diba1013** [comment]
1. We thought some context might be nice, for locating the element within the page 
2. We just implemented is for a single web element, as (for now) the list can be easily stuffed into multiple checks and the PageObject needs to expose its element to be able to check. 
For the List and the PageObject, we still need to decide on how we want to support/implement it.
3. I assume that you use `Unbreakable` and `Autochecking` with `RecheckWebImpl`. Both need context for being able to find the element in question, which this single element cannot provide.

We will look into updating the documentation accordingly.

**Comment by javydreamercsw** [comment]
WebElement interface provides all that information once initialized;

- Point getLocation()

- Dimension getSize()

- Rectangle getRect()

- String getAttribute(String name)

Maybe even [this](https://stackoverflow.com/a/27611777/198108) could give more information.

**Comment by beatngu13** [comment]
@javydreamercsw check out the new beta:

https://github.com/retest/recheck-web/releases/tag/v1.5.0-beta.2

It now only creates a screenshot of the element you are checking.

**Comment by roesslerj** [comment]
> Also it seems to only work with RecheckImpl() and not with RecheckWebImpl(). Took me a while to figure out why it didn't work on my project at work.

Will be addressed, thanks for pointing that out.

**Comment by roesslerj** [comment]
> Started giving it a try and noticed that the AutocheckingRecheckDriver constructor changed. There seems to no longer be a way to set a naming strategy. In my case I used to pass a custom AutocheckingCheckNamingStrategy and that's no longer possible it seems.

Will be fixed, thanks for pointing it out. Would you mind creating individual issues? Makes it easier to track ;-) Thanks!

**Comment by javydreamercsw** [comment]
Just tried the change to only screen shot the WebElement and works perfectly! Just had to add a NameStrategy like suggested.

But got a report loading exception:

```
Exception loading a Report

java.lang.ClassCastException: class java.lang.Float cannot be cast to class de.retest.recheck.persistence.Persistable (java.lang.Float is in module java.base of loader 'bootstrap'; de.retest.recheck.persistence.Persistable is in unnamed module of loader 'app')
	at de.retest.recheck.persistence.bin.KryoPersistence.load(SourceFile:52)
	at de.retest.recheck.persistence.PersistenceFactory$DynamicLoadPersistenceProxy.load(SourceFile:85)
	at de.retest.report.LoadTestReportFromInputFlow.a(SourceFile:83)
	at de.retest.report.LoadTestReportFromInputFlow.a(SourceFile:135)
	at de.retest.report.LoadTestReportFromInputFlow.a(SourceFile:75)
	at de.retest.report.LoadTestReportFromInputFlow.a(SourceFile:53)
	at de.retest.gui.review.workers.LoadTestReportWorker.a(SourceFile:69)
	at de.retest.gui.review.workers.LoadTestReportWorker.doInBackground(SourceFile:39)
	at java.desktop/javax.swing.SwingWorker$1.call(SwingWorker.java:304)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.desktop/javax.swing.SwingWorker.run(SwingWorker.java:343)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

```

**Comment by javydreamercsw** [comment]
Seems to work fine as I'm seeing the screenshots of the element themselves but ran into the issues above.

**Comment by roesslerj** [comment]
Yeah, sorry ... we changed the binary report format, and it's not upwards compatible: You need a GUI of 1.5.0+ to open the reports. But it's downwards compatible, meaning that the GUI can open older reports...

**Comment by roesslerj** [comment]
You can download the GUI (also beta) from here: https://assets.retest.org/releases/review/review-1.5.0-beta.2-win.zip

**Comment by javydreamercsw** [comment]
Was unable to use that with the following error:
 ```
java -jar review.jar 
16:42:14.115 [main] INFO de.retest.util.MainUtil - retest rocks!
16:42:14.137 [main] INFO de.retest.util.MainUtil - Launching retest version 1.5.0-beta.2 via Java(TM) SE Runtime Environment 1.8.0_151-b12 (Oracle Corporation) on Mac OS X 10.14.6 (x86_64).
16:42:14.336 [main] WARN de.retest.recheck.logging.LogUtil - No logback.xml found or configured! Please review your setting of property 'logback.configurationFile'.
16:42:14.339 [main] INFO de.retest.logging.LogFileFinder - Getting path to retest log-file not implemented for logger of class ch.qos.logback.classic.LoggerContext.
16:42:14.339 [main] INFO de.retest.util.MainUtil - Writing log file to 'null'.
16:42:14.346 [main] WARN de.retest.configuration.RetestWorkspace - No properties file found, no user properties loaded! Specify valid system property 'de.retest.configFile'.
16:42:14.347 [main] WARN de.retest.configuration.RetestWorkspace - Properties de.retest.configFile and de.retest.workDirectory are not set, assuming default ('null').
16:42:14.463 [retest-check-internet-connection] ERROR de.retest.configuration.PropertyVariableResolver - The property 'de.retest.workDirectory' was not defined but needed by the property 'de.retest.vmArgs'.
16:42:14.467 [retest-check-internet-connection] ERROR de.retest.configuration.PropertyVariableResolver - The property 'pom.parent.version' was not defined but needed by the property 'de.retest.version'.
16:42:15.177 [retest-check-internet-connection] INFO de.retest.internet.ReTestConnection - Working Internet connection: RequestResult [responseCode=200, stream=org.apache.commons.io.input.AutoCloseInputStream@6753fe0f, error=null]
16:42:15.980 [main] ERROR de.retest.gui.helper.ReTestErrorHandler - Uncaugt exception in thread Thread[main,5,main]
java.lang.NullPointerException: null
	at de.retest.configuration.Configuration.getRetestWorkspace(SourceFile:72)
	at de.retest.license.LicenseFileReader.<init>(SourceFile:47)
	at de.retest.license.LicenseManager.g(SourceFile:68)
	at de.retest.license.LicenseManager.<init>(SourceFile:63)
	at de.retest.license.LicenseManager.a(SourceFile:37)
	at de.retest.util.MainUtil.a(SourceFile:43)
	at de.retest.gui.ReTestGui.main(SourceFile:41)

Exception: java.lang.NullPointerException thrown from the UncaughtExceptionHandler in thread "main"
```

**Comment by roesslerj** [comment]
Interesting. Thanks for finding that before the release... will fix asap.

**Comment by roesslerj** [comment]
Fixed in the next beta.

**Comment by roesslerj** [comment]
But to make it work until the next beta, could you use either the .bat, .exe (are you on Windows?) or .sh files to start review?

**Comment by javydreamercsw** [comment]
I'm on Mac. I'll try when I get to the office.

**Comment by javydreamercsw** [comment]
<img width="291" alt="Screen Shot 2019-09-17 at 7 39 26 AM" src="https://user-images.githubusercontent.com/2817343/65042359-4ddbaf80-d91e-11e9-93e9-e7ccc6d56ded.png">
No script files either.

**Comment by javydreamercsw** [comment]
![Screen Shot 2019-09-17 at 7 42 07 AM](https://user-images.githubusercontent.com/2817343/65042546-ac089280-d91e-11e9-829b-39bdbfddad2d.png)
No accept/ignore/apply for differences either.

**Comment by roesslerj** [comment]
The scripting error is my bad. I was thinking you are on windows, so I linked the windows zip file, which doesn't contain a shell script. The all platforms download would have been https://assets.retest.org/releases/review/review-1.5.0-beta.2.zip.

Whether you see apply/ignore checkboxes depends on whether you are in a project context (Golden Masters available) and whether your license is still valid. We'll change that in the next release (1.6), such that without license you don't have checkboxes and without Golden Masters checkboxes are disabled. Since your license will expire in 8 days, review cannot locate the Golden Master files. From where did you load the report? Did that change to before?

**Comment by javydreamercsw** [comment]
Loaded from the same place (project target folder). Was not aware of the licensing part of the review tool.

**Comment by javydreamercsw** [comment]
Maybe the space in the path? That's most likely different. I opened another report in the same place but the name has no space and the check boxes show up.

**Comment by roesslerj** [comment]
That could be the case. Thanks for pointing that out.

**Comment by beatngu13** [comment]
[Newest beta](https://github.com/retest/recheck-web/releases/tag/v1.5.0-beta.4) should fix the [regression mentioned here](https://github.com/retest/recheck-web/issues/352#issuecomment-533126977).

**Comment by javydreamercsw** [comment]
Everything looking fine on beta 4. Just a few log errors I want to get to your attention:

```
08:36:22.370 ERROR [pool-5-thread-2] [de.retest.recheck.ui.image.ImageUtils] - X value of image to cut 595 is outside of bigger image with width 580!
 08:36:22.370 ERROR [pool-5-thread-1] [de.retest.recheck.ui.image.ImageUtils] - X value of image to cut 595 is outside of bigger image with width 290!
 08:36:22.371 ERROR [pool-5-thread-1] [de.retest.recheck.ui.image.ImageUtils] - Cannot create mark at x value 595 with in image of width 290.
 08:36:22.372 ERROR [pool-5-thread-2] [de.retest.recheck.ui.image.ImageUtils] - Cannot create mark at x value 595 with in image of width 580.
```
I suspect those are tied to the web element screenshots. No functionality seems affected but just in case.

I can send the report file for inspection to a private email if needed.

**Comment by beatngu13** [comment]
Implemented in [v1.5.0](https://github.com/retest/recheck-web/releases/tag/v1.5.0), [documentation](https://github.com/retest/docs/pull/44) on its way.
